### PR TITLE
[DEV] fix layerwise torch_dist checkpointing fails due to empty rank

### PR DESCRIPTION
**Pull request by FDecaYed**
# What does this PR do ?
Fix torch_dist checkpointing format fail when there are empty groups due to layerwise sharding.
cherry-picked into master muon PR here https://github.com/NVIDIA/Megatron-LM/pull/2241 for main

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 1]
```diff
@@ -225,6 +225,22 @@ def step(self):  # type: ignore[no-untyped-def]
...
+    # TODO(deyuf): need to improve dist checkpointing design to properly handle this
+    # fp32_from_fp16_params is list, each sub list could be empty if group is empty
+    # this breaks dist checkpointing assumption since extract_sharded_base drop list structure
+    # for now, we convert it to dict with index as key and convert back in load_state_dict
+    def load_state_dict(self, state_dict):
+        if len(self.chained_optimizers) == 1:
+            wrapped_state_dict = {1: state_dict}
+        else:
+            wrapped_state_dict = state_dict
+        for sd in wrapped_state_dict.values():
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py:240]: Suggestion: Consider adding logging string.  infor would suffice.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 2]
```diff
@@ -246,27 +262,45 @@ def sharded_state_dict(
...
+                local_params = group.pop('params')
+                # save whether this group is empty, so we can use non-empty rank for metadata
+                group['params'] = bool(local_params.unwrap())
+                all_rank_groups = [None for _ in range(get_pg_size(self.pg_collection.dp_cp))]
+                torch.distributed.all_gather_object(
+                    all_rank_groups, group, self.pg_collection.dp_cp
+                )
+                nonempty_rank_group = next((g for g in all_rank_groups if g['params']), None)
+                if nonempty_rank_group is None:
+                    raise ValueError(
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py]: This is more like RuntimeError rather than value error

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 3]
```diff
@@ -246,27 +262,45 @@ def sharded_state_dict(
...
+            for i, group in enumerate(sd['optimizer']['param_groups']):
+                # keep local param tensor so we only gather metadata
+                local_params = group.pop('params')
+                # save whether this group is empty, so we can use non-empty rank for metadata
+                group['params'] = bool(local_params.unwrap())
+                all_rank_groups = [None for _ in range(get_pg_size(self.pg_collection.dp_cp))]
+                torch.distributed.all_gather_object(
+                    all_rank_groups, group, self.pg_collection.dp_cp
+                )
+                nonempty_rank_group = next((g for g in all_rank_groups if g['params']), None)
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py]: This line is hard to read, try not to use next.
If it is too hard to be a one liner, make it a loop and append to nonempty_rank_group is completely fine.

**Review by skyw — COMMENTED** [review]
Could we also add unit test ? single transformer layer with DP8 should do.

**Comment by skyw** [comment]
Often line tested more cases and they all seemed to be working.

two warnings remain, I think they are normal?

```bash
WARNING:megatron.core.utils:make_tp_sharded_tensor_for_checkpoint received extra kwargs: ['allow_shape_mismatch']

WARNING:megatron.core.dist_checkpointing.optimizer:Sharded parameters mapping is empty. It means tensors in model state dict do not correspond to tensors in optimizer parameters map. Make sure to call state_dict with `keep_vars=True`.
```

**Comment by FDecaYed** [comment]
> ```shell
> WARNING:megatron.core.utils:make_tp_sharded_tensor_for_checkpoint received extra kwargs: ['allow_shape_mismatch']
> ```

this seems to be a issue that's in TOT base branch and not related to this PR
> ```shell
> WARNING:megatron.core.dist_checkpointing.optimizer:Sharded parameters mapping is empty. It means tensors in model state dict do not correspond to tensors in optimizer parameters map. Make sure to call state_dict with `keep_vars=True`.
> ```
This is expected since we're re-using megatron's method for mapping optim state to param. but in our case, some rank will have empty group. Do you think we should update the warning and mention 'if you're not using layerwise'?

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 4]
```diff
@@ -246,27 +262,39 @@ def sharded_state_dict(
...
-            step = next(
-                (group['step'] for group in sd['optimizer']['param_groups'] if 'step' in group),
-                None,
-            )
-            if step is not None:
-                for group in sd['optimizer']['param_groups']:
-                    group['step'] = step
-
+                sd['fp32_from_fp16_params'] = {
+                    k: v for k, v in enumerate(sd['fp32_from_fp16_params'])
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py]: This is a bit confusing. I would use i instead of k here because it is not taking key from dict.item() iterator where  k,v is mostly commonly used.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 5]
```diff
@@ -246,27 +262,39 @@ def sharded_state_dict(
...
+                sd['optimizer']['state'] = LocalNonpersistentObject(sd['optimizer']['state'])
+            # group keys(e.g. 'step') might be missing or not updated
+            for i, group in enumerate(sd['optimizer']['param_groups']):
+                # keep local param tensor so we only gather metadata
+                local_params = group.pop('params')
+                # save whether this group is empty, so we can use non-empty rank for metadata
+                group['params'] = bool(local_params.unwrap())
+                all_rank_groups = [None for _ in range(torch.distributed.get_world_size())]
+                torch.distributed.all_gather_object(all_rank_groups, group)
+                nonempty_rank_group = next((g for g in all_rank_groups if g['params']), group)
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py:300]: This is hard to read. Suggest to rewrite without next

**Review by skyw — APPROVED** [review]
some minor suggestions, otherwise LGTM.

Should have a follow up PR to add test protecting previous failed cases.

File: megatron/core/optimizer/layer_wise_optimizer.py (hunk 5 above)
Comment (FDecaYed) [reply megatron/core/optimizer/layer_wise_optimizer.py:300]: rewrite without next probably takes 5 line of code. I added a comment to explain this line and feel leave it this way is ok
